### PR TITLE
Add optional metadata grabber for television: ttvdbv4

### PR DESCRIPTION
**Pull request by rcrdnalor**
Add optional metadata grabber for television: ttvdbv4

- [x ] contribution does not duplicate one of our [existing pull requests](https://github.com/MythTV/mythtv/pulls)
- [x ] contribution is in a branch rebased against [master](https://github.com/MythTV/mythtv)
- [x ] code compiles successfully without errors
- [x ] code follows the [MythTV Coding Standards](https://www.mythtv.org/wiki/Coding_Standards)
- [x ] documentation added/updated/removed where necessary
- [x ] commits are logically organised and have [good commit messages](https://chris.beams.io/posts/git-commit)

Motivation:
TheTVDB.com moves to API 4.x and deprecates the API 3.x.

See
issue https://github.com/MythTV/mythtv/issues/301
https://support.https://support.thetvdb.com//
https://api.swaggerhub.com/apis/thetvdb/tvdb-api_v_4/4.3.8
latest specification as listed in
https://api.swaggerhub.com/apis/thetvdb/tvdb-api_v_4/

This commit adds an alternative metadata grabber for TV-Series.
The ttvdb4 metadata grabber is up to date with the latest 
api definitions of thetvdb.com and follows the intentions in
https://www.mythtv.org/wiki/MythTV_Universal_Metadata_Format

To enable this grabber, run in mythtv frontend
"Setup" -> "Artwork and Data Sources"
and select 'TheTVDatabaseV4' for the default TelevisionGrabber.

Once done, you can check this global setting with mysql:
`
[mythconverg]> select * from settings where value like '%grabber%';

+-------------------------+--------------------------------+--------------+
| value                      | data                                | hostname |
+-------------------------+--------------------------------+--------------+
| TelevisionGrabber  | metadata/Television/ttvdb4.py  | NULL     |
| MovieGrabber       | metadata/Movie/tmdb3.py        | NULL     |
+--------------------+--------------------------------+----------+
`
Any further metadata update will then use these grabbers.

Note: The ttvdb4 grabber is compatible to python 3.6+.

Features:
This grabber uses a file named 'ttvdb4.ini' which is
placed in the '$MYTHCONFDIR' folder which defaults to 
'/home/mythtv/.mythtv'. This file allows you to customize
the search behavior, similar to the options mentioned in 
https://www.mythtv.org/wiki/Ttvdb.py
An additional search order by languages was added as requested by #180 .
Look at the file located at '$MYTHCONFDIR/ttvdbv4.ini'
and edit it accordingly.
The first language to seach is allways the language passed via
command line, e.g.: 'ttvdb4.py -l de ....'
You can verify the used settings by passing '--debug' to the seach.
This grabber detects duplicate entries for episode names, like "Episode1"
for multiple seasons. This was requested by developer jpoet.
The '--jsondebug' switch lists the raw data retrieved from thetvdb.com.

Caveats:
Due to the added capabilities, the search for data may take longer than
3 minutes. In this case, one can find an entry in mythfrontend.log like
"Child PID 27965 killed with Terminated".

Usage:
Use 'ttvdb4.py --help' for details.

Note: The option 'iso-date-time' is currently not implemented.

Developer Note:
Albeit the provided TheTVdbv4 specification claims to be compatible to 
the OpenApiSpecification oas 3.0.0,
there are some bugs in the provided spec and in the delivered data structures.
Refs: 
  https://www.openapis.org/
  https://api.swaggerhub.com/apis/thetvdb/tvdb-api_v_4/4.3.8
This hinders the usage of generic tools for an API-client like 
  https://github.com/OpenAPITools/openapi-generator
or
  https://fastapi.tiangolo.com/
Known issues are reported in the commit messages of this grabber.
For those reasons, the api for TheTVdbv4 is parsed by a handcrafted tool
'get_api_v4.py' and manual overrides have been added.

### CODE REVIEW

File: mythtv/bindings/python/ttvdbv4/definitions.py [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+        self.aliases = _get_list(data, 'aliases')
+        self.companies = _get_list(data, 'companies')
+        self.companyType = data.get('companyType', '')                       # string
+        self.country = data.get('country', '')                               # string
+        self.director = data.get('director', '')                             # string
+        self.extendedTitle = data.get('extendedTitle', '')                   # string
+        self.genres = _get_list(data, 'genres')
+        self.id = data.get('id', '')                                         # string
+        self.image_url = data.get('image_url', '')                           # string
+### XXX self.imageUrl = data.get('imageUrl', '')                             # string
```
Comment (mspieth) [inline mythtv/bindings/python/ttvdbv4/definitions.py:503]: should these comments be deleted?

File: mythtv/bindings/python/ttvdbv4/myth4ttvdbv4.py [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+        except:
+            pass
+
+        if self.ShowPeople:
+            # add characters: see getAllPeopleTypes
+            ## type: 3: Actor, 2: Writer, 1: Director, 4: Guest Star, 5: Crew, 6: Creator
+            ## 7: Producer, 8: Showrunner, 9: Musical Guest, 10: Host, 11: Executive Producer
+
+            # characters of type 'Actor' are sorted in ascending order
+            actors = [c for c in ser_x.characters if c.type == 3]
```
Comment (mspieth) [inline mythtv/bindings/python/ttvdbv4/myth4ttvdbv4.py:355]: would it be better to add an enum class for the enum?

File: mythtv/programs/scripts/metadata/Television/ttvdb4_doctests [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+>>> main()
+<?xml version='1.0' encoding='UTF-8'?>
+<metadata>
+  <item>
+    <title>Chernobyl</title>
+    <description>Chernobyl dramatizes the story of the 1986 nuclear accident — one of the worst man-made catastrophes in history — and the sacrifices made to save Europe from unimaginable disaster.</description>
+    <inetref>360893</inetref>
+    <homepage>https://www.hbo.com/chernobyl/season-1</homepage>
+    <language>en</language>
+    <userrating>8.700000</userrating>
```
Comment (mspieth) [inline mythtv/programs/scripts/metadata/Television/ttvdb4_doctests:24]: consider wildcarding anything that could be brittle. e.g. userrating is something that could change

File: mythtv/bindings/python/ttvdbv4/myth4ttvdbv4.py [hunk 4]
```diff
@@ -207,8 +207,7 @@ def __init__(self, **kwargs):
 
         self.session.headers.update({'Accept': 'application/json',
                                      'Accept-Language': self.language,
-                                     ### XXX ToDo define User Agent
-                                     # 'User-Agent': 'mythtv ttvdb4 grabber %s' % __version__,
+                                     'User-Agent': 'mythtv.org ttvdb v.4 grabber',
```
Comment (mspieth) [inline mythtv/bindings/python/ttvdbv4/myth4ttvdbv4.py:210]: should there be a . in v.4? v4 seems better to me.

**Review by mspieth — APPROVED** [review]
LGTM

**Review by mspieth — APPROVED** [review]

File: mythtv/bindings/python/ttvdbv4/definitions.py [hunk 5]
```diff
@@ -1,5 +1,11 @@
 # -*- coding: UTF-8 -*-
 
+# ---------------------------------------
```
Comment (mspieth) [inline mythtv/bindings/python/ttvdbv4/definitions.py:3]: Should there be some sort of mythtv copyright info or yourname copyright. gpl/bsd/apache/mit?

**Review by mspieth — APPROVED** [review]

**Comment by rcrdnalor** [comment]
Add thetvdb.com API v4 metadata grabber for television

The implementation of the TV grabber is based on following documents:
https://github.com/thetvdb/v4-api/blob/main/docs/swagger.yml (v 4.4.0)

This commit adds an alternative metadata grabber for TV-Series.
The ttvdb4 metadata grabber is up to date with the latest
api definitions of thetvdb.com and follows the intentions in
https://www.mythtv.org/wiki/MythTV_Universal_Metadata_Format

To enable this grabber, run in mythtv frontend
"Setup" -> "Artwork and Data Sources"
and select 'TheTVDatabaseV4' for the default TelevisionGrabber.

**Pull request merged**